### PR TITLE
Emit constant tensors as global constants

### PR DESCRIPTION
**Pull request by doru1004**
This patch allows the transformation of constant tensors specified as ONNXConstantOp operations as global constants in LLVM.

This patch performs the following transformations:

```
%0 = "onnx.Constant"() {value = dense<[[0.0, 0.0], [1.0, 1.1], [2.0, 2.1]]> : tensor<3x2xf32>} : () -> tensor<*xf32>
```

========= ONNXConstantOpLowering =======>

```
%0 = "krnl.global"() {name = "constant_0", shape = [3, 2], value = dense<[[0.000000e+00, 0.000000e+00], [1.000000e+00, 1.100000e+00], [2.000000e+00, 2.100000e+00]]> : tensor<3x2xf32>} : () -> memref<3x2xf32>
```

========= KrnlGlobalOpLowering =======>

```
llvm.mlir.global internal constant @constant_0(dense<[[0.000000e+00, 0.000000e+00], [1.000000e+00, 1.100000e+00], [2.000000e+00, 2.100000e+00]]> : tensor<3x2xf32>) : !llvm<"[3 x [2 x float]]">
```

The global constant then needs to be transformed into an LLVM MemRef such that the value can be passed to other operations that expect a MemRef type.

First, the constant array is copied into a local alloca (in line with what happens for C/C++ constant arrays):
```
    %5 = llvm.mlir.constant(1 : i64) : !llvm.i64
    %6 = llvm.alloca %5 x !llvm<"[3 x [2 x float]]"> : (!llvm.i64) -> !llvm<"[3 x [2 x float]]*">
    %7 = llvm.bitcast %6 : !llvm<"[3 x [2 x float]]*"> to !llvm<"i8*">
    %8 = llvm.mlir.addressof @constant_0 : !llvm<"[3 x [2 x float]]*">
    %9 = llvm.bitcast %8 : !llvm<"[3 x [2 x float]]*"> to !llvm<"i8*">
    %10 = llvm.mlir.constant(4 : i64) : !llvm.i64
    %11 = llvm.mlir.constant(6 : i64) : !llvm.i64
    %12 = llvm.mul %10, %11 : !llvm.i64
    %13 = llvm.sext %12 : !llvm.i64 to !llvm.i64
    %14 = llvm.mlir.constant(0 : i1) : !llvm.i1
    %15 = llvm.call @llvm.memcpy.p0i8.p0i8.i64(%7, %9, %13, %14) : (!llvm<"i8*">, !llvm<"i8*">, !llvm.i64, !llvm.i1) -> !llvm.void
```

The output of the ConstantOp needs to produce a MemRef so a local MemRef is emitted as follows:
```
    %16 = llvm.mlir.undef : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
```

The local MemRef is then filled in with the allocated and aligned data pointers as well as the spatial sizes and strides:
```
    %17 = llvm.bitcast %6 : !llvm<"[3 x [2 x float]]*"> to !llvm<"float*">
    %18 = llvm.insertvalue %17, %16[0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %19 = llvm.insertvalue %17, %18[1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %20 = llvm.mlir.constant(0 : index) : !llvm.i64
    %21 = llvm.insertvalue %20, %19[2] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %22 = llvm.mlir.constant(3 : index) : !llvm.i64
    %23 = llvm.insertvalue %22, %21[3, 0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %24 = llvm.mlir.constant(1 : index) : !llvm.i64
    %25 = llvm.insertvalue %24, %23[4, 0] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %26 = llvm.mlir.constant(2 : index) : !llvm.i64
    %27 = llvm.insertvalue %26, %25[3, 1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
    %28 = llvm.mlir.constant(1 : index) : !llvm.i64
    %29 = llvm.insertvalue %28, %27[4, 1] : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
```

Note:

The handling of the constant arrays matches the handling of constant arrays in C/C++.
Compiling the LLVM code above with -S shows that the constant tensors emitted this way end up in the read-only data section.

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
After explanations from Doru, the code in the lower segment are filling in the fields that characterize the data structure.

```
!llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
```

and where the constant memory is copied into the alloca-ed memory. This is the way memref works and is handled properly by LLVM.

**Comment by tungld** [comment]
@doru1004 I checked this patch with MNIST and ResNet50. Here are summary:
(E2E test includes 3 steps: 1. produce byte codes => 2. produce object file => 3. link and run)
- [MNIST 1](https://github.com/onnx/models/tree/master/vision/classification/mnist): E2E test ran fast, taking under 1s and producing a correct output.
- [MNIST 2](https://s3.us.cloud-object-storage.appdomain.cloud/dlcompiler/model/mnist.onnx):   
    - `onnx-mlir --EmitLLVMBC` took *12 seconds* to produce byte codes.
    - E2E test got segmentation fault at step 3.
- [ResNet50](https://github.com/onnx/models/tree/master/vision/classification/resnet/resnet50) + PR #38 (for AveragePool)
    - `onnx-mlir --EmitLLVMBC` took about *6 minutes*.
    - E2E test: took 13 minutes for steps 1 and 2 , got segmentation fault at step 3.

In summary, this patch reduces compilation time significantly. Not sure segmentation fault is related to this patch or not.

File: src/Transform/LowerToLLVM.cpp [hunk 1]
```diff
@@ -60,6 +62,164 @@ static size_t getRankFromMemRefType(LLVM::LLVMType memRefTy) {
...
+    Value isVolatile = rewriter.create<LLVM::ConstantOp>(
+        loc, LLVM::LLVMType::getInt1Ty(llvmDialect),
+        rewriter.getIntegerAttr(rewriter.getIntegerType(1), 0));
+    //  - Copy constant data into the alloca.
+    auto memcpyRef = getOrInsertMemcpy(rewriter, module, llvmDialect);
+    rewriter.create<CallOp>(
+        loc, memcpyRef, LLVM::LLVMType::getVoidTy(llvmDialect),
+        ArrayRef<Value>({int8PtrAlloc, i8PtrGlobal, int64Size, isVolatile}));
+
+    // Create the llvm.mlir.undef corresponding to the MemRef.
```
Comment (tungld) [inline src/Transform/LowerToLLVM.cpp:190]: Lines 190--211 can be replaced by
```mlir
auto llvmConstantElementType = constantElementType.cast<LLVM::LLVMType>();                                                                                             
Value typedAlloc = rewriter.create<LLVM::BitcastOp>(                                                                                                                   
    loc, llvmConstantElementType.getPointerTo(), alloc);                                                                                                               
auto llvmMemRef = MemRefDescriptor::fromStaticShape(                                                                                                                   
    rewriter, loc, typeConverter, memRefTy, typedAlloc); 
```
which helps create a MemRefType.

Comment (doru1004) [reply src/Transform/LowerToLLVM.cpp]: I updated this. Thanks for pointing this out.
